### PR TITLE
Update lronacpho to use LROC's newer photometric model and parameters file

### PR DESCRIPTION
**Pull request by victoronline**
## Description
This update needs to set the default photometric model (LROC_Empirical) to the newer model (2019) and look for the default parameters file in the data/mission/calibration directory. If the old parameter file is provided, the old algorithm(2014) will be used.

Default behavior:
Should use a newer photometric model as default with the newer parameters file (NAC_PHO_LROC_Empirical.0003.pvl).

Optional behavior:
If another pvl is provided with and doesn't specify the algorithm version (NAC_PHO_LROC_Empirical.0001.pvl), then the old photometric model will be used.

If the 2014 version is specified, such as in NAC_PHO_LROC_Empirical.0002.pvl, then the old photometric model will be used.

## Related Issue
[#4512 ](https://github.com/USGS-Astrogeology/ISIS3/issues/4512)

## Motivation and Context
LROC has created a newer photometric model requiring updated parameters. 

## How Has This Been Tested?
This has been tested by the science team comparing photometric products using both the old and newer models and validated the results.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

**Comment by victoronline** [comment]
What do we need to do to get this moving forward? Can I work on the missing pieces, like gtest, or do I have to wait for the review to be made?

**Comment by victoronline** [comment]
What do we need to do to get this moving forward? Can I work on the missing pieces, like gtest, or do I have to wait for the review to be made?

File: isis/src/lro/apps/lronacpho/LROCEmpirical.cpp
Comment (jessemapel) [inline isis/src/lro/apps/lronacpho/LROCEmpirical.cpp:199]: We usually give a list of valid options in error messages like this. Maybe:

"Algorithm version in PVL file not recognized [" + IString(parms.algoVersion) + "]. Supported version are [0, 2014, 2019]."

File: isis/src/lro/apps/lronacpho/LROCEmpirical.cpp
Comment (jessemapel) [inline isis/src/lro/apps/lronacpho/LROCEmpirical.cpp]: why is this using `toString(0.0)` for the last argument and not just "0.0"?

File: isis/src/lro/apps/lronacpho/LROCEmpirical.cpp
Comment (jessemapel) [inline isis/src/lro/apps/lronacpho/LROCEmpirical.cpp:267]: This looks like an indentation error. Indents should be 2 spaces

File: isis/src/lro/apps/lronacpho/LROCEmpirical.cpp
Comment (jessemapel) [inline isis/src/lro/apps/lronacpho/LROCEmpirical.cpp]: Is the `QString("0")` required instead of `"0"`?

File: isis/src/lro/apps/lronacpho/main.cpp
Comment (jessemapel) [inline isis/src/lro/apps/lronacpho/main.cpp]: Why the change here? It's technically still true as ProcessByLine inherits from ProcessByBrick

File: isis/src/lro/apps/lronacpho/main.cpp
Comment (jessemapel) [inline isis/src/lro/apps/lronacpho/main.cpp]: You can embed this logic in the app XML, see how hical specifies this:

https://github.com/USGS-Astrogeology/ISIS3/blob/dev/isis/src/mro/apps/hical/hical.xml#L1485

File: isis/src/lro/apps/lronacpho/main.cpp
Comment (jessemapel) [inline isis/src/lro/apps/lronacpho/main.cpp]: If you swap to how hical handles this you don't need to worry about this. If you want to stick with this setup, then you can just use FileName::expanded to expand out `$lro/calibration` to the LRO calibration directory.

File: isis/src/lro/apps/lronacpho/LROCEmpirical.cpp
Comment (victoronline) [reply isis/src/lro/apps/lronacpho/LROCEmpirical.cpp:199]: Updated with suggested changes.

File: isis/src/lro/apps/lronacpho/LROCEmpirical.cpp
Comment (victoronline) [reply isis/src/lro/apps/lronacpho/LROCEmpirical.cpp:267]: Fixed indentation

File: isis/src/lro/apps/lronacpho/LROCEmpirical.cpp
Comment (victoronline) [reply isis/src/lro/apps/lronacpho/LROCEmpirical.cpp]: I'm not sure why that was there. Must have been a remnant of old code or maybe copy/paste error. Thanks!

File: isis/src/lro/apps/lronacpho/LROCEmpirical.cpp
Comment (victoronline) [reply isis/src/lro/apps/lronacpho/LROCEmpirical.cpp]: Fixed

File: isis/src/lro/apps/lronacpho/main.cpp
Comment (victoronline) [reply isis/src/lro/apps/lronacpho/main.cpp]: We were trying different ways to speed up processing in different scenarios and left this comment in erroneously. Thanks for the catch.

File: isis/src/lro/apps/lronacpho/main.cpp
Comment (victoronline) [reply isis/src/lro/apps/lronacpho/main.cpp]: Thanks Jesse. Updated with suggested changes.

File: isis/src/lro/apps/lronacpho/main.cpp
Comment (victoronline) [reply isis/src/lro/apps/lronacpho/main.cpp]: Swapped it out. Thanks!

**Comment by victoronline** [comment]
@jessemapel I think I hit all items on your review. Let me know if you need anything else before proceeding.

**Comment by victoronline** [comment]
If there aren't any issues, is there a chance to get this into the next sprint?

**Comment by victoronline** [comment]
@jessemapel Just making sure this doesn't get lost. Please let me know if you still need anything from me.

**Comment by jessemapel** [comment]
This looks good now. I'm not sure how to handle updating the tests on this. For the moment, can you e-mail me or attach a zip of a test image with the validated output?

**Comment by victoronline** [comment]
Jesse, I'll send that to you here shortly. Thanks.

**Comment by jessemapel** [comment]
For testing on this, I'd like to see one test that uses the old parameter file, this likely already exists, and one test that uses the new parameter file. The tests should check that the proper parameters are actually being used and that the report is correct.

**Comment by victoronline** [comment]
@scsides, 

This is the error I'm getting when changing the app to be callable. I'll take a closer look later tonight but if something jumps out, please let me know. It's probably an easy typo.

`In file included from /ser/estore/scratch/isis/development/git_dev/ISIS3/isis/src/lro/apps/lronacpho/lronacpho.h:6:0,
                 from /ser/estore/scratch/isis/development/git_dev/ISIS3/isis/src/lro/apps/lronacpho/main.cpp:2:
inc/Isis.h:9:2: error: #error *****Isis.h MUST be included before any other files!*****
 #error *****Isis.h MUST be included before any other files!*****
  ^~~~~
In file included from /ser/estore/scratch/isis/development/git_dev/ISIS3/isis/src/lro/apps/lronacpho/lronacpho.h:6:0,
                 from /ser/estore/scratch/isis/development/git_dev/ISIS3/isis/src/lro/apps/lronacpho/main.cpp:2:
inc/Isis.h: In function ‘std::map<QString, void*> GuiHelpers()’:
inc/Isis.h:71:27: error: redefinition of ‘std::map<QString, void*> GuiHelpers()’
 std::map<QString, void *> GuiHelpers() {
                           ^~~~~~~~~~
In file included from /ser/estore/scratch/isis/development/git_dev/ISIS3/isis/src/lro/apps/lronacpho/main.cpp:1:0:
inc/Isis.h:71:27: note: ‘std::map<QString, void*> GuiHelpers()’ previously defined here
 std::map<QString, void *> GuiHelpers() {
                           ^~~~~~~~~~`

**Comment by Kelvinrr** [comment]
@victoronline I can look at the code if need be, the fact that there seems to be a merge issue here (look at files changed) makes it kinda hard. But basically that error is telling you that ISIS.h is not included where it needs to be. It should only be at the very top of the `main.cpp` where `IsisMain` is located and not in `<app>.cpp` where the app-as-funcs are being defined. Common mistake if you copy pasted code over and forgot to delete `#include "Isis.h"` at the top.

**Comment by scsides** [comment]
@victoronline: In the version I pulled down earlier, lronacpho.cpp was included, but I don't see it now.  It had a semicolon at the end of the #include for Isis.h, but Isis.h only needs to be included in the main. Isis.h actually houses the C++ int main(argc, argv){} function and should not be include anywhere else.

**Comment by scsides** [comment]
lronacpho.cpp has a { on line 6. Should be ;

From what I'm seeing, the Isis.h error was masking compile/syntax errors like the one above

**Comment by victoronline** [comment]
I'm fixing the merge errors. I'll upload it again here shortly.

**Comment by jlaura** [comment]
@Kelvinrr Are you able to take a look at this today and get a review back to @victoronline?

@victoronline This has merge conflicts that are going to need to be fixed before it is mergeable. Let us know if you have questions about how to handle those.

Thanks!

**Comment by victoronline** [comment]
@jlaura Merge conflicts resolved.

**Comment by victoronline** [comment]
Should I also check the test cub into the data directory?

**Comment by jlaura** [comment]
@victoronline how large is the test cub? We might want to crop not down first if possible.

@Kelvinrr @jessemapel what is the best method to get test data from an external source? Email if small enough and we will check in to the test data on PR merge?

File: isis/tests/FunctionalTestsLronacpho.cpp
Comment (Kelvinrr) [inline isis/tests/FunctionalTestsLronacpho.cpp:37]: Assuming this image is small enough (like single digit MB), you can just check it in as part of this PR. It seems you already cropped it so I image its good as is.

Comment (victoronline) [reply isis/tests/FunctionalTestsLronacpho.cpp:37]: It is not a single-digit MB. It is 8.4 MB

Comment (victoronline) [reply isis/tests/FunctionalTestsLronacpho.cpp:37]: I'll crop down to 1 MB and add to PR

Comment (victoronline) [reply isis/tests/FunctionalTestsLronacpho.cpp:37]: Added test cube

**Comment by victoronline** [comment]
@jlaura I didn't see your comment before I added the test cube. I can still email if needed.

**Comment by jlaura** [comment]
@victoronline Nope, less than a MB and in the data area like @Kelvinrr said is great! 

@Kelvinrr is going to give the updated code a review so that we can get it merged since a number of commits have come in since the original review.

File: isis/src/lro/apps/lronacpho/LROCEmpirical.cpp
Comment (Kelvinrr) [inline isis/src/lro/apps/lronacpho/LROCEmpirical.cpp]: I noticed Jesse suggested changing the `toString` call with the string literal `"0.0"`, and that you made the change, was this reverted somehow? Or do you need to explicitly cast to a `QString`?

File: isis/src/lro/apps/lronacpho/LROCEmpirical.cpp
Comment (Kelvinrr) [inline isis/src/lro/apps/lronacpho/LROCEmpirical.cpp]: same string question here.

File: isis/src/lro/apps/lronacpho/lronacpho.h
Comment (Kelvinrr) [inline isis/src/lro/apps/lronacpho/lronacpho.h]: kinda minor, but might as well change `appLog` here to `log` for consistency with how they're defined in the rest of the `lronacpho` signatures.

File: isis/tests/FunctionalTestsLronacpho.cpp
Comment (Kelvinrr) [inline isis/tests/FunctionalTestsLronacpho.cpp:141]: Anything particular about this DN that makes it important to check or is it a random sample?

**Review by Kelvinrr — COMMENTED** [review]
Just a few questions, overall looks fine to me. The only other change I would make before approving is adding a line to CHANGELOG.MD under "Unreleased", basically a short description of the change with a link to the relevant issue similar to the history logs. After that I'll approve assuming no other objections from anyone else and we can get this merged.

**Comment by jlaura** [comment]
@victoronline Do you need anything else from us on this right now or are you good with making the minor adjustments based on the re-review? Then we can get this merged and start on the next PR.

**Comment by victoronline** [comment]
@jlaura  I should be able to address the items above shortly. Appreciate the note.

File: isis/tests/FunctionalTestsLronacpho.cpp
Comment (victoronline) [reply isis/tests/FunctionalTestsLronacpho.cpp:141]: This is just a random sample compared to the same output using fx equation the science team is using manually. I would've included testing against fx so any random pixel value could be checked but fx isn't yet callable. This should be updated once it is in case a cube that doesn't have (300,300) pixel.

File: isis/src/lro/apps/lronacpho/lronacpho.h
Comment (victoronline) [reply isis/src/lro/apps/lronacpho/lronacpho.h]: Updated in code.

File: isis/src/lro/apps/lronacpho/LROCEmpirical.cpp
Comment (victoronline) [reply isis/src/lro/apps/lronacpho/LROCEmpirical.cpp]: This must have gotten reverted when I corrected the merge issue. Thanks. Fixed in code.

File: isis/src/lro/apps/lronacpho/LROCEmpirical.cpp
Comment (victoronline) [reply isis/src/lro/apps/lronacpho/LROCEmpirical.cpp]: This must have gotten reverted when I corrected the merge issue. Thanks. Fixed in code.

**Comment by victoronline** [comment]
I don't know the etiquette regarding who does the "Resolve conversation" here so I left it up to the reviewer to resolve.

**Comment by victoronline** [comment]
@Kelvinrr @jlaura Do you all need anything else from me to get this merged?

**Review by Kelvinrr — APPROVED** [review]
@victoronline looks good!  🎉

**Pull request merged**
